test/e2e/es/volume_test.go

```diff
@@ -191,7 +191,13 @@ func TestVolumeExpansion(t *testing.T) {
 						return err
 					}
 					patchStorageSize(&es, resizedStorage)
-					return k.Client.Update(context.Background(), &es)
+					err := k.Client.Update(context.Background(), &es)
+					if err != nil {
+						return err
+					}
+					// update the expected nodeSets in the builder because the ES resource has just been modified by an external mechanism
+					b.WithExpectedNodeSets(es.Spec.NodeSets...)
```

Comment:
> Essentially, it's to make *elasticsearch.Builder implement test.Builder

Could we not just rely on 2 builders here ? One to create the initial deployment and a second one to describe the expected state ?
Something along those lines:

```golang
	// Create a copy of the builder with the expected storage resources.
	scaledUpStorage := b.DeepCopy()
	for i := range scaledUpStorage.Elasticsearch.Spec.NodeSets {
		scaledUpStorage.Elasticsearch.Spec.NodeSets[i].VolumeClaimTemplates[0].Spec.Resources.Requests[corev1.ResourceStorage] = resizedStorage
	}

	test.Sequence(nil, func(k *test.K8sClient) test.StepList {
		return test.StepList{
			{
				Name: "Update the Elasticsearch spec with increased storage requests",
				Test: test.Eventually(func() error {
					var es esv1.Elasticsearch
					if err := k.Client.Get(context.Background(), k8s.ExtractNamespacedName(&b.Elasticsearch), &es); err != nil {
						return err
					}
					patchStorageSize(&es, resizedStorage)
					return k.Client.Update(context.Background(), &es)
				}),
			},
			{
				Name: "PVCs should eventually be resized",
				Test: test.Eventually(func() error {
					for _, pvcName := range pvcNames {
						var pvc corev1.PersistentVolumeClaim
						if err := k.Client.Get(context.Background(), types.NamespacedName{Namespace: b.Elasticsearch.Namespace, Name: pvcName}, &pvc); err != nil {
							return err
						}
						reportedStorage := pvc.Status.Capacity.Storage()
						if reportedStorage == nil {
							return fmt.Errorf("no storage size reported in %s status", pvcName)
						}
						if !reportedStorage.Equal(resizedStorage) {
							return fmt.Errorf("expected resized capacity %s but got %s", resizedStorage.String(), reportedStorage.String())
						}
					}
					return nil
				}),
			},
			{
				Name: "StatefulSets should eventually be recreated with the right storage size",
				Test: test.Eventually(func() error {
					for _, ssetName := range []string{masterSset, dataSset} {
						var sset appsv1.StatefulSet
						if err := k.Client.Get(context.Background(), types.NamespacedName{Namespace: b.Elasticsearch.Namespace, Name: ssetName}, &sset); err != nil {
							return err
						}
						if !sset.Spec.VolumeClaimTemplates[0].Spec.Resources.Requests.Storage().Equal(resizedStorage) {
							return fmt.Errorf("StatefulSet %s has not been recreated with storage size %s", ssetName, resizedStorage.String())
						}
					}
					return nil
				}),
			},
			// re-run all the regular checks
		}.WithSteps(test.CheckTestSteps(scaledUpStorage, k))
	}, b).RunSequential(t)
```

Reply:
It's much smarter indeed.